src/linearmixedmodel.jl

```diff
@@ -296,23 +291,28 @@ function condVar(m::LinearMixedModel{T}) where {T}
     Array{T,3}[reshape(abs2.(ll ./ Ld) .* varest(m), (1, 1, length(Ld)))]
 end
 
-function createAL(allterms::Vector{Union{AbstractReMat{T},FeMat{T}}}) where {T}
-    k = length(allterms)
-    sz = [isa(t, ReMat) ? size(t, 2) : rank(t) for t in allterms]
-    A = AbstractMatrix{T}[]
-    L = AbstractMatrix{T}[]
-    for i = 1:k
-        for j = 1:i
-            Lij = densify(allterms[i]' * allterms[j])
-            push!(L, Lij)
-            push!(A, deepcopy(isa(Lij, BlockedSparse) ? Lij.cscmat : Lij))
+function pushALblock!(A, L, blk)
+    push!(L, blk)
+    push!(A, deepcopy(isa(blk, BlockedSparse) ? blk.cscmat : blk))
+end
+function createAL(reterms::Vector{AbstractReMat{T}}, Xy::FeMat{T}) where {T}
+    k = length(reterms)
+    vlen = kchoose2(k+1)
+    A = sizehint!(AbstractMatrix{T}[], vlen)
+    L = sizehint!(AbstractMatrix{T}[], vlen)
+    for i in eachindex(reterms)
+        for j in 1:i
+            pushALblock!(A, L, densify(reterms[i]' * reterms[j]))
         end
     end
-    nretrm = sum(Base.Fix2(isa, ReMat), allterms)
-    for i = 2:nretrm      # check for fill-in due to non-nested grouping factors
-        ci = allterms[i]
-        for j = 1:(i-1)
-            cj = allterms[j]
+    for j in eachindex(reterms)   # can't fold this into the previous loop b/c block order
+        pushALblock!(A, L, densify(Xy' * reterms[j]))
+    end
+    pushALblock!(A, L, densify(Xy'Xy))
```

Comment:
I'm very curious whether we any sparse FE design matrix actually gets stored as sparse after the product here....